### ISSUE TITLE
B::Deparse fails to correctly dereference the sub named '&'

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#10545]
Migrated from [rt.perl.org#77172](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=77172) (status was 'resolved')

Searchable as RT77172$

**Comment by p5pRT** [comment]
### From alex@pearanalytics.com

#### Created by alex@pearanalytics.com



B&#8203;::Deparse can produce an invalid syntactical transformation in the case of
the
subroutine & &&#8203;:

&nbsp;    $ perl \-MO=Deparse \-e 'sub foo \{\} foo & & & foo'
&nbsp;    sub foo \{

&nbsp;    \}
&nbsp;    foo && & foo\(\);
&nbsp;    \-e syntax OK

This transformation is wrong; the doubly\-deparsed result is wildly different
from what B&#8203;::Concise thinks is going on\.

&nbsp;    $ perl \-MO=Deparse \-e 'sub foo \{\} foo & & & foo' | perl \-MO=Deparse
&nbsp;    \-e syntax OK
&nbsp;    sub foo \{

&nbsp;    \}
&nbsp;    &foo\(\) if foo ;

The syntax tree as reported by B&#8203;::Concise&#8203;:

&nbsp;    $ perl \-MO=Concise\,\-exec \-e 'sub foo \{\} foo & & & foo'
&nbsp;    1  \<0> enter
&nbsp;    2  \<;> nextstate\(main 2 \-e&#8203;:1\) v&#8203;:\{
&nbsp;    3  \<0> pushmark s
&nbsp;    4  \<0> pushmark s
&nbsp;    5  \<\#> gv\[\*&\] s
&nbsp;    6  \<1> entersub\[t3\] sK/TARG\,AMPER\,1
&nbsp;    7  \<0> pushmark s
&nbsp;    8  \<\#> gv\[\*foo\] s
&nbsp;    9  \<1> entersub\[t5\] sKS/TARG\,1
&nbsp;    a  \<2> bit\_and\[t6\] sKM
&nbsp;    b  \<\#> gv\[\*foo\] s
&nbsp;    c  \<1> entersub\[t7\] vKS/TARG\,1
&nbsp;    d  \<@&#8203;> leave\[1 ref\] vKP/REFC

I admit that the above case is kind of perverse\, as the bug is reproducible
in
only three characters&#8203;:

&nbsp;    $ perl \-MO=Deparse \-e '& &' | perl \-c
&nbsp;    \-e syntax OK
&nbsp;    syntax error at \- line 1\, near "&&"
&nbsp;    \- had compilation errors\.

It seems like the simplest fix would be to add whitespace when deparsing
calls
to & &\.

<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=low
    module=B::Deparse

Site configuration information for perl 5.10.1:

Configured by Debian Project at Fri Apr 23 08:18:41 UTC 2010.

Summary of my perl5 (revision 5 version 10 subversion 1) configuration:

  Platform:
    osname=linux, osvers=2.6.24-27-server,
archname=x86_64-linux-gnu-thread-multi
    uname='linux crested 2.6.24-27-server #1 smp fri mar 12 01:23:09 utc
2010 x86_64 gnulinux '
    config_args='-Dusethreads -Duselargefiles -Dccflags=-DDEBIAN
-Dcccdlflags=-fPIC -Darchname=x86_64-linux-gnu -Dprefix=/usr
-Dprivlib=/usr/share/perl/5.10 -Darchlib=/usr/lib/perl/5.10
-Dvendorprefix=/usr -Dvendorlib=/usr/share/perl5 -Dvendorarch=/usr/lib/perl5
-Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl/5.10.1
-Dsitearch=/usr/local/lib/perl/5.10.1 -Dman1dir=/usr/share/man/man1
-Dman3dir=/usr/share/man/man3 -Dsiteman1dir=/usr/local/man/man1
-Dsiteman3dir=/usr/local/man/man3 -Dman1ext=1 -Dman3ext=3perl
-Dpager=/usr/bin/sensible-pager -Uafs -Ud_csh -Ud_ualarm -Uusesfio -Uusenm
-DDEBUGGING=-g -Doptimize=-O2 -Duseshrplib -Dlibperl=libperl.so.5.10.1
-Dd_dosuid -des'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN
-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include
-D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2 -g',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fno-strict-aliasing -pipe
-fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.4.3', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t',
lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib /lib64 /usr/lib64
    libs=-lgdbm -lgdbm_compat -ldb -ldl -lm -lpthread -lc -lcrypt
    perllibs=-ldl -lm -lpthread -lc -lcrypt
    libc=/lib/libc-2.11.1.so, so=so, useshrplib=true,
libperl=libperl.so.5.10.1
    gnulibc_version='2.11.1'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -g -L/usr/local/lib
-fstack-protector'

Locally applied patches:



@INC for perl 5.10.1:
    /etc/perl
    /usr/local/lib/perl/5.10.1
    /usr/local/share/perl/5.10.1
    /usr/lib/perl5
    /usr/share/perl5
    /usr/lib/perl/5.10
    /usr/share/perl/5.10
    /usr/local/lib/site_perl
    .


Environment for perl 5.10.1:
    HOME=/home/alex
    LANG=en_US.utf8
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)

 PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

Fixed by 1b38d78\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'new' to 'resolved'
